transports/quic/src/transport.rs

```diff
@@ -223,7 +288,38 @@ impl<P: Provider> Transport for GenTransport<P> {
         }
 
         if let Poll::Ready(Some(ev)) = self.listeners.poll_next_unpin(cx) {
-            return Poll::Ready(ev);
+            match ev {
+                TransportEvent::Incoming {
+                    listener_id,
+                    upgrade,
+                    local_addr,
+                    send_back_addr,
+                } => {
+                    let socket_addr =
+                        multiaddr_to_socketaddr(&send_back_addr, self.support_draft_29)
+                            .unwrap()
+                            .0;
+
+                    if let Some(sender) = self.hole_punch_attempts.remove(&socket_addr) {
+                        if let Err(upgrade) = sender.send(upgrade) {
+                            return Poll::Ready(TransportEvent::Incoming {
+                                listener_id,
+                                upgrade,
+                                local_addr,
+                                send_back_addr,
+                            });
+                        }
+                    } else {
+                        return Poll::Ready(TransportEvent::Incoming {
+                            listener_id,
+                            upgrade,
+                            local_addr,
+                            send_back_addr,
+                        });
+                    }
```

Comment:
```suggestion
                    if let Some(sender) = self.hole_punch_attempts.remove(&socket_addr) {
                        match sender.send(upgrade) {
                             Ok(()) => continue,
                             Err(timed_out_holepunch) => {
                                 upgrade = timed_out_holepunch;
                             }
                        }
                    }

                    return Poll::Ready(TransportEvent::Incoming {
                            listener_id,
                            upgrade,
                            local_addr,
                            send_back_addr,
                        });
```

Does this work? Naming subject to improvements :)

This also highlights that we actually have to use a loop here (around the listeners), otherwise we return `Pending` despite not having exhausted the listeners.